### PR TITLE
feat(deploy): create dedicated lockfile

### PR DESCRIPTION
**Pull request by KSXGitHub**
Closes https://github.com/pnpm/pnpm/issues/8778

### CODE REVIEW

**Comment by haines** [comment]
Will this also fix https://github.com/pnpm/pnpm/issues/5315?

File: releasing/plugin-commands-deploy/test/deploy.test.ts [hunk 1]
```diff
@@ -77,6 +97,513 @@ test('deploy', async () => {
...
+    {
+      name: 'project-0',
+      version: '0.0.0',
+      private: true,
+      dependencies: {
+        'project-0': 'workspace:*',
+        // NOTE: there is an inexplicable bug in which 'renamed-workspace' and 'renamed-linked' cannot exist simultaneously
+        // TODO: fix this bug
+        'renamed-workspace': 'workspace:project-0@*',
+        // 'renamed-linked': 'link:.',
```
Comment (KSXGitHub) [inline releasing/plugin-commands-deploy/test/deploy.test.ts]: I cannot reproduce this bug with `pd`. I don't know how to fix this.

File: releasing/plugin-commands-deploy/src/deploy.ts [hunk 2]
```diff
@@ -134,3 +147,78 @@ async function copyProject (src: string, dest: string, opts: { includeOnlyPackag
...
+  }
+
+  if (!lockfileDir) {
+    return 'opts.lockfileDir is undefined. Falling back to installing without a lockfile.'
+  }
+
+  if (!workspaceDir) {
+    return 'opts.workspaceDir is undefined. Falling back to installing without a lockfile.'
+  }
+
```
Comment (zkochan) [inline releasing/plugin-commands-deploy/src/deploy.ts]: Are these possible scenarios? Maybe just throw errors.

**Review by zkochan — COMMENTED** [review]
overall looks good. I will review it more and test it later.

File: releasing/plugin-commands-deploy/src/deploy.ts (hunk 2 above)
Comment (KSXGitHub) [reply releasing/plugin-commands-deploy/src/deploy.ts]: https://github.com/pnpm/pnpm/pull/8828/commits/0e3f2cd92d9fbe46a0e3e1df018e0c4c8a79563b

File: releasing/plugin-commands-deploy/src/createDeployFiles.ts [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    return path.resolve(projectRootDirRealPath, targetPath)
+  }
+
+  if (spec.startsWith('file:')) {
+    const targetPath = spec.slice('file:'.length)
+    return path.resolve(lockfileDir, targetPath)
+  }
+
+  return undefined
+}
```
Comment (KSXGitHub) [inline releasing/plugin-commands-deploy/src/createDeployFiles.ts]: I tried to find an existing function in the `resolver` directory, but none of them can fit nicely.

Also, this whole function can be converted to a single regex. Should we do that?

File: releasing/plugin-commands-deploy/test/deploy.test.ts (hunk 1 above)
Comment (KSXGitHub) [reply releasing/plugin-commands-deploy/test/deploy.test.ts]: Fixed by https://github.com/pnpm/pnpm/pull/8828/commits/ba79dff7278976e27c47947ccdad19fde9d5d513

**Comment by zkochan** [comment]
With these changes deploy currently fails if there is are `pnpm.overrides`, `pnpm.packageExtensions`, or `pnpm.patchedDependencies`.

**Comment by KSXGitHub** [comment]
@zkochan There's one test left that I don't know how to write. I wanted to test a lockfile with peer dependencies suffix in the `ResolvedDependency` (similar to the bug you caught when you tried to deploy `pnpm@v8`), but I don't know how to write a package structure to make these suffixes appear.

**Comment by zkochan** [comment]
There are some package mocks with peer dependencies. You can see them in these tests for instance: https://github.com/pnpm/pnpm/blob/main/pkg-manager/core/test/install/peerDependencies.ts

**Comment by KSXGitHub** [comment]
> There are some package mocks with peer dependencies. You can see them in these tests for instance: https://github.com/pnpm/pnpm/blob/main/pkg-manager/core/test/install/peerDependencies.ts

Peer dependencies from registry packages are not what I'm worrying about. It's the peer dependencies from the workspace packages. I have been failing to create a lockfile with syntax `project-1@file:project-1(peer1@spec1)(peer2@spec2)`.

File: releasing/plugin-commands-deploy/src/createDeployFiles.ts [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+    output[key] = createFileUrlDepPath(depRealPath, opts.allProjects)
+  }
+
+  return output
+}
+
+function resolveLinkOrFile (spec: string, opts: Pick<ConvertOptions, 'lockfileDir' | 'projectRootDirRealPath'>): string | undefined {
+  // try parsing `spec` as `spec(peers)`
+  const hasPeers = /^(?<spec>[^()]+)(?<peers>\(.+\))$/.exec(spec)
```
Comment (zkochan) [inline releasing/plugin-commands-deploy/src/createDeployFiles.ts:264]: you don't need to write new parsers. There are already parsers for depPath in `@pnpm/dependency-path`

**Comment by zkochan** [comment]
> Peer dependencies from registry packages are not what I'm worrying about. It's the peer dependencies from the workspace packages. I have been failing to create a lockfile with syntax project-1@file:project-1(peer1@spec1)(peer2@spec2).

That is because [dedupe-injected-deps](https://pnpm.io/npmrc#dedupe-injected-deps) is `true`. It is easier to do if you set that setting to `false` and the workspace projects have peer dependencies.

**Comment by zkochan** [comment]
Here's an example of a workspace with peer prefix with `file:` dep: https://github.com/pnpm/pnpm/blob/d86015b09c8d4090c2f23ec80a4360e38b0ab22f/pkg-manager/core/test/install/injectLocalPackages.ts#L1782

This one is with `dedupeInjectedDeps: true`.

**Comment by KSXGitHub** [comment]
> That is because [dedupe-injected-deps](https://pnpm.io/npmrc#dedupe-injected-deps) is `true`. It is easier to do if you set that setting to `false` and the workspace projects have peer dependencies.

I see. I will try it.

File: releasing/plugin-commands-deploy/src/createDeployFiles.ts (hunk 4 above)
Comment (KSXGitHub) [reply releasing/plugin-commands-deploy/src/createDeployFiles.ts:264]: I can't make it work. I have published the change to a branch named [try-using-dependency-path](https://github.com/pnpm/pnpm/tree/try-using-dependency-path).

latest commit: https://github.com/pnpm/pnpm/commit/0dce3be50463879ce81abbb76571f7b476cbd4bb

diff: https://github.com/pnpm/pnpm/compare/deploy-prod-fails-due-to-missing-dev-8778...try-using-dependency-path

Comment (KSXGitHub) [reply releasing/plugin-commands-deploy/src/createDeployFiles.ts:264]: The test regarding peer suffix fails. Attempting to deploy `@pnpm/headless` from `pnpm@v8` also fails.

Comment (KSXGitHub) [reply releasing/plugin-commands-deploy/src/createDeployFiles.ts:264]: `parseDepPath` doesn't work either: https://github.com/pnpm/pnpm/tree/try-using-dependency-path-2

Comment (zkochan) [reply releasing/plugin-commands-deploy/src/createDeployFiles.ts:264]: > The test regarding peer suffix fails. Attempting to deploy @pnpm/headless from pnpm@v8 also fails.

I will need a few hours to look into it, which is problematic now. I will try to look into it in the next couple of days.

**Review by zkochan — APPROVED** [review]

**Pull request merged**